expertsystem/data.py

```diff
@@ -343,31 +343,29 @@ def compute_isospin_projection(  # pylint: disable=too-many-arguments
         )
 
 
-class ParticleCollection(abc.Mapping):
-    """Safe, `dict`-like collection of `.Particle` instances."""
+class ParticleCollection(abc.MutableSet):
+    """Searchable collection of immutable `.Particle` instances."""
 
     def __init__(self, particles: Optional[Iterable[Particle]] = None) -> None:
         self.__particles: Dict[str, Particle] = dict()
+        self.__pid_to_name: Dict[int, str] = dict()
         if particles is not None:
-            if not isinstance(particles, (list, set, tuple)):
-                raise ValueError(
-                    f"Cannot construct a {self.__class__.__name__} "
-                    f"from a {particles.__class__.__name__}"
-                )
-            self.__particles.update(
-                {
-                    particle.name: particle
-                    for particle in particles
-                    if isinstance(particle, Particle)
-                }
-            )
-
-    def __contains__(self, particle_name: object) -> bool:
-        return particle_name in self.__particles
+            self.update(particles)
+
+    def __contains__(self, instance: object) -> bool:
+        if isinstance(instance, str):
+            return instance in self.__particles
+        if isinstance(instance, Particle):
+            return instance in self.__particles.values()
+        if isinstance(instance, int):
+            return instance in {p.pid for p in self}
```

Comment:
this would change to `instance in self.__pid_to_name`

Reply:
Oops, yes 14cf964f